File: 2025/rust/src/bin/day01.rs
use aoc2025::{Day, run};
use std::str::FromStr;
use thiserror::Error;

#[derive(Error, Debug)]
enum Error {
    #[error("invalid rotation")]
    InvalidRotation,
}

#[derive(Debug)]
enum Rotation {
    L(i32),
    R(i32),
}

impl FromStr for Rotation {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (dir, num) = s.split_at(1);
        Ok(match (dir, num.parse()) {
            ("L", Ok(n)) => Rotation::L(n),
            ("R", Ok(n)) => Rotation::R(n),
            _ => return Err(Error::InvalidRotation),
        })
    }
}

struct Input(Vec<Rotation>);

impl FromStr for Input {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        s.lines()
            .map(str::parse)
            .collect::<Result<_, _>>()
            .map(Input)
    }
}

struct Solution;

impl Solution {
    fn rotations(input: &Input) -> impl Iterator<Item = (i32, i32)> {
        input.0.iter().scan(50, |state, rot| match rot {
            Rotation::L(n) => {
                // Turning left decreases the position
                let p = (100 - *state) % 100;
                *state = (*state - n).rem_euclid(100);
                Some((*state, (p + n) / 100))
            }
            Rotation::R(n) => {
                let p = *state;
                *state = (*state + n).rem_euclid(100);
                Some((*state, (p + n) / 100))
            }
        })
    }
}

impl Day for Solution {
    type Input = Input;
    type Error = Error;
    type Answer1 = usize;
    type Answer2 = i32;

    fn part1(input: &Self::Input) -> Self::Answer1 {
        Solution::rotations(input).filter(|(p, _)| *p == 0).count()
    }
    fn part2(input: &Self::Input) -> Self::Answer2 {
        Solution::rotations(input).map(|(_, c)| c).sum()
    }
}

fn main() -> Result<(), aoc2025::Error<Error>> {
    run::<Solution>()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn example_input() -> Result<Input, Error> {
        r"
L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
"
        .trim_start()
        .parse()
    }

    #[test]
    fn part1() -> Result<(), Error> {
        assert_eq!(Solution::part1(&example_input()?), 3);
        Ok(())
    }

    #[test]
    fn part2() -> Result<(), Error> {
        assert_eq!(Solution::part2(&example_input()?), 6);
        Ok(())
    }
}

File: 2025/ocaml/bin/day01.ml
open Aoc2025
open Base

module Solution : Aoc.Day = struct
  type rotation = Left | Right
  type input = (rotation * int) list

  let parse channel =
    let parse_line s =
      let rot =
        match String.get s 0 with
        | 'L' -> Left
        | 'R' -> Right
        | _ -> failwith "invalid input"
      in
      (rot, String.subo s ~pos:1 |> Int.of_string)
    in
    In_channel.input_lines channel |> List.map ~f:parse_line

  let count_zeros lst =
    let f acc (rot, n) =
      let next =
        (match rot with Left -> acc + n | Right -> acc - n)
        |> Fn.flip ( % ) 100
      in
      (next, next)
    in
    List.folding_map lst ~init:50 ~f |> List.count ~f:(( = ) 0)

  let part1 input = count_zeros input |> Answer.of_int

  let part2 input =
    List.map input ~f:(fun (rot, n) -> List.init n ~f:(fun _ -> (rot, 1)))
    |> List.concat |> count_zeros |> Answer.of_int
end

let () =
  let part =
    Sys.get_argv () |> Array.last |> function
    | "part1" -> Aoc.Part1
    | "part2" -> Aoc.Part2
    | _ -> Aoc.Both
  in
  Aoc.run ~part (module Solution) In_channel.stdin
